src/poetry/utils/exporter.py

```diff
@@ -70,21 +70,23 @@ def _export_requirements_txt(
         content = ""
         dependency_lines = set()
 
-        for package, groups in itertools.groupby(
-            self._poetry.locker.get_project_dependency_packages(
-                project_requires=self._poetry.package.all_requires,
-                dev=dev,
-                extras=extras,
-            ),
-            lambda dependency_package: dependency_package.package,
+        # Get project dependencies, and add the project-wide marker to them.
```

Comment:
Why should the project level marker be required here? Isin't this causing more noise without any realy functional benifit as the python compatibility refers to the project and not it's dependencies.

Reply:
this is necessary to get the right answers!  we are now more careful in the main walk about selecting packages that are compatible with python markers - https://github.com/dimbleby/poetry/blob/2d61c3475d8746d6ff1d6f3e1cdfb0070815b468/src/poetry/packages/locker.py#L221 - so we need to pass in a suitably restricted initial marker or we might not find any solution at all.

Reply:
Since the exporter should rely on locked versions only, I do not see why we won't find a solution as the locker is aware of the project constraints. The reason why I am not a big fan of this is because the exports now look like this.

```
foo==1.0.0 ; python_version >= "2.7" and python_version < "2.8" or python_version >= "3.4" and python_version < "4.0"
```

While I agree that this is more exact, this is not really ideal. I can see the community not being happe about it as we already have had issues pile up when markers became a bit verbose in earlier iterations.

Reply:
I understand.  But the line of code that I pointed at really is an essential correctness fix: and once you are committed to that you have to make sure that you start with a sufficiently narrow range of python versions, else that line will reject packages with python requirements that are too broad.

I'm open to other approaches if they can be found, I just don't currently see it.  I encourage you to experiment if you have other ideas in mind - the new `test_exporter_doesnt_confuse_repeated_packages` testcase work turned out to uncover lots of interesting bugs...

I suspect that most of us who export `requirements.txt` don't much care about the cosmetics any more than we care about the contents of `poetry.lock` - and anyway we're willing to put up with a little extra noise if that's the trade for getting the right answer.

Reply:
Understandable. The way I might approach this is to pass `python_constraint=root_package.python_constraint` into `Locker.get_project_dependency_packages()` and then into the other methods. And do something like `package.python_constraint.allows_all(dependency.python_constraint.intersect(python_constraint))` in `Locker.__get_locked_package()`. 

This allows the flexibility to fetch dependencies given a python constraint and lets us keep the current behaviour.

We can discuss the change in behaviour in a separate PR and focus on the refactor and fixes here.

> I suspect that most of us who export requirements.txt don't much care about the cosmetics

Personally, I would agree. However, in certain cases readability has been indicated as a preference in some issues. I just want to decouple that discussion from the functional changes here.

Reply:
That's do-able.  However it results in requirements.txt files that are ugly in a different way - and that are also inaccurate.

Consider a project whose python constraint is `^3.8`, and which happens to use some package A whose marker includes a broader constraint than this: say `^2.7 or ^3.6`.

Then under your proposal A and its children carry this maker through the search, and so that's what shows up in the exported `requirements.txt`.  

That's both ugly _and_ wrong!

In the MR as it stands, this broader marker gets merged with the project constraint, and so what is exported is just `>= 3.8 and <4`.

I acknowledge that the new output is noisier than it used to be: but I really would argue that it is now correct when it previously wasn't.  It seems completely reasonable to me that the exported requirements.txt should have the same constraints as the project being exported.

Reply:
I'd add that as a result of recent work to improve simplifications of marker union and intersection, we are in general getting much more compact markers out than we used to.  

That doesn't change the question of whether including the project marker is in general desirable or not; but if the main worry is only cosmetic then it does greatly soften the blow.

Reply:
actually for some combination of amusement and clarification... `poetry export` on poetry itself is kind of entertaining.

[this](https://gist.github.com/dimbleby/b5d2ae5a3d269a865a08bd9730ccc601) is what poetry 1.1.3 does.  And [this](https://gist.github.com/dimbleby/a0c465efe2d8e3c38dd779a459346918) is what this MR gets us.

After is both neater and more correct.  It's hard to understand what 1.1.3's:
```
zipp==3.7.0; python_version >= "3.7" and python_version < "3.8" and (python_version >= "3.6" and python_full_version < "3.0.0" and python_version < "3.8" or python_full_version >= "3.5.0" and python_version < "3.8" and python_version >= "3.6") or python_version < "3.8"
```
really means, but it certainly insists on `<"3.8"`, which is wrong.  (`importlib-metadata` is installed for all versions, so `zipp` must follow).

This is mostly just for fun, of course the master branch benefits from marker simplifications regardless of how we resolve this particular sub-question.

Reply:
actually, for completeness, here's what we get following your idea of hiding the project python constraint during the search: https://gist.github.com/dimbleby/d0954dbce65411af1904f98957242060

it demonstrates very clearly the problem I noted - I don't think it's an improvement

Reply:
> That's both ugly and wrong!

I would argue that the changes proposed in this PR only makes the edge case look better, and makes the majority cases look unnecessarily noisy. However, it is not wrong. As per the lock file, the applied constraint is `^2.7 or ^3.6` for `A`. And hence it is dutifully dumping the marker as is, or atleast that is what I would expect.

> That doesn't change the question of whether including the project marker is in general desirable or not; but if the main worry is only cosmetic then it does greatly soften the blow.

At present, for me, there are the following open questions:
1. Cosmetic - this is not so important. Although I personally prefer the less noisy version as it is (a) easier to reason about than having all lines with markers view inspecting, (b) maps to real-world use cases of `requirements.txt`, (c) it is redundant for almost all use cases of the export.
2. Extensibility -  In the near future two changes are expected that are related to this. The first is de-coupling of the export command into its own plugin. And the second, is the use of the `Locker` for various cases like that of inspecting the lockfile (build, third-party tools etc).

Let's assume that we say we do not care about (1), I get the feeling you feel strongly about it :wink:. Let's keep the project constraint in as per your current changes. We can always deal with it later if there is push back from the community.

Hwoever, for (2), I would like to have the logic for "intersecting" root project python constraint be moved into `get_project_dependency_packages` (however that is done, passing the project or passing in the required python constraint). This is because otherwise we end up having to re-implement this logic for any use case that inspects dependencies from the lockfile in order to not fail.

Reply:
> I would argue that the changes proposed in this PR only makes the edge case look better, and makes the majority cases look unnecessarily noisy. 

Per the example of poetry itself in https://github.com/python-poetry/poetry/pull/5156#discussion_r833719416: your "edge case" may actually be the majority case!

Anyway, I've added a commit which I think is what you're asking for, let me know if I've misunderstood.

Reply:
No, that looks about right. Thank you.